File: samples/python/copilotstudio-skill/src/agent.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from os import environ

from dotenv import load_dotenv
from microsoft_agents.hosting.core import (
    Authorization,
    AgentApplication,
    TurnState,
    TurnContext,
    MemoryStorage,
)
from microsoft_agents.activity import load_configuration_from_env
from microsoft_agents.hosting.aiohttp import CloudAdapter
from microsoft_agents.authentication.msal import MsalConnectionManager

load_dotenv()
agents_sdk_config = load_configuration_from_env(environ)

VERSION = agents_sdk_config.get("version", "unknown")
STORAGE = MemoryStorage()
CONNECTION_MANAGER = MsalConnectionManager(**agents_sdk_config)
ADAPTER = CloudAdapter(connection_manager=CONNECTION_MANAGER)
AUTHORIZATION = Authorization(STORAGE, CONNECTION_MANAGER, **agents_sdk_config)

AGENT_APP = AgentApplication[TurnState](
    storage=STORAGE, adapter=ADAPTER, authorization=AUTHORIZATION, **agents_sdk_config
)


@AGENT_APP.conversation_update("membersAdded")
async def on_members_added(context: TurnContext, _state: TurnState):
    welcome_text = f"Hello from echo bot, running on version {VERSION}"
    await context.send_activity(welcome_text)


@AGENT_APP.activity("message")
async def on_message(context: TurnContext, _state: TurnState):
    text = context.activity.text
    reply_text = f"Echo: {text}"
    await context.send_activity(reply_text)
    if "version" in text.lower():
        await context.send_activity(f"Running on version {VERSION}")

File: samples/nodejs/copilotstudio-skill/src/agent.ts
import express from 'express'
import path from 'path'
import { AgentApplication, MessageFactory, CloudAdapter, authorizeJWT, getAuthConfigWithDefaults } from '@microsoft/agents-hosting'
import pjson from '@microsoft/agents-hosting/package.json'

export const skillAgent = new AgentApplication()

skillAgent.onConversationUpdate('membersAdded', async (context) => {
  const welcomeText = `Hello from echo bot, running on version ${pjson.version}`
  await context.sendActivity(MessageFactory.text(welcomeText, welcomeText))
})

skillAgent.onActivity('message', async (context) => {
  const text = context.activity.text
  const replyText = `Echo: ${text}`
  await context.sendActivity(MessageFactory.text(replyText, replyText))
  if (text?.includes('version')) {
    await context.sendActivity(MessageFactory.text('Running on version ' + pjson.version, 'Running on version ' + pjson.version))
  }
})

skillAgent.onActivity('endOfConversation', async (context) => {
  // Handle PVASkillImport ping and endOfConversation from Copilot Studio
})

const authConfig = getAuthConfigWithDefaults()
const adapter = skillAgent.adapter || new CloudAdapter()

const server = express()
server.use(express.json())

// Serve static files BEFORE auth so /manifest.json is publicly accessible
server.use(express.static(path.join(__dirname, '..', 'public')))

// Auth only on the messages endpoint
server.post('/api/messages', authorizeJWT(authConfig), (req, res) =>
  adapter.process(req, res, (context) => skillAgent.run(context))
)

const port = process.env.PORT || 3978
server.listen(port, () => {
  console.log(`\nServer listening to port ${port} on sdk ${pjson.version} for appId ${authConfig.clientId} debug ${process.env.DEBUG}`)
})
